src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -558,537 +631,447 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
+        part->info.min_block = block_number;
+        part->info.max_block = block_number;
+        part->info.level = 0;
+        part->info.mutation = 0;
+
+        part->setName(part->getNewName(part->info));
+        retry_context.actual_part_name = part->name;
+
+        /// Prepare transaction to ZooKeeper
+        /// It will simultaneously add information about the part to all the necessary places in ZooKeeper and remove block_number_lock.
+        Coordination::Requests ops;
+
+        get_logs_ops(ops);
+
+        /// Deletes the information that the block number is used for writing.
+        size_t block_unlock_op_idx = ops.size();
+        block_number_lock->getUnlockOp(ops);
+
+        get_quorum_ops(ops);
+
+        size_t shared_lock_ops_id_begin = ops.size();
+        storage.getLockSharedDataOps(*part, zookeeper, /*replace_zero_copy_lock*/ false, {}, ops);
+        size_t shared_lock_op_id_end = ops.size();
+
+        storage.getCommitPartOps(ops, part, block_id_path);
 
         /// It's important to create it outside of lock scope because
         /// otherwise it can lock parts in destructor and deadlock is possible.
         MergeTreeData::Transaction transaction(storage, NO_TRANSACTION_RAW); /// If you can not add a part to ZK, we'll remove it back from the working set.
-        bool renamed = false;
-
         try
         {
             auto lock = storage.lockParts();
-            renamed = storage.renameTempPartAndAdd(part, transaction, lock);
+            storage.renameTempPartAndAdd(part, transaction, lock);
         }
         catch (const Exception & e)
         {
-            if (e.code() != ErrorCodes::DUPLICATE_DATA_PART && e.code() != ErrorCodes::PART_IS_TEMPORARILY_LOCKED)
-                throw;
-        }
-
-        if (!renamed)
-        {
-            if (is_already_existing_part)
+            if (e.code() == ErrorCodes::DUPLICATE_DATA_PART || e.code() == ErrorCodes::PART_IS_TEMPORARILY_LOCKED)
             {
-                LOG_INFO(log, "Part {} is duplicate and it is already written by concurrent request or fetched; ignoring it.", part->name);
-                return;
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                                "Part with name {} is already written by concurrent request."
+                                " It should not happen for non-duplicate data parts because unique names are assigned for them. It's a bug",
+                                part->name);
             }
-            else
-                throw Exception(ErrorCodes::LOGICAL_ERROR, "Part with name {} is already written by concurrent request."
-                    " It should not happen for non-duplicate data parts because unique names are assigned for them. It's a bug",
-                    part->name);
-        }
-
-        auto rename_part_to_temporary = [&temporary_part_relative_path, &transaction, &part]()
-        {
-            transaction.rollbackPartsToTemporaryState();
```

Comment:
This breaks retries when the commit to zookeeper fails, since the part will not be in a temporary state.

```
[azat.local] 2023.12.09 17:18:02.833553 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Debug> executeQuery: (from [::1]:54026) insert into data_r1 values  (stage: Complete)
[azat.local] 2023.12.09 17:18:02.833673 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> ContextAccess (default): Access granted: INSERT(key) ON default.data_r1
[azat.local] 2023.12.09 17:18:02.834525 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77): Trying to reserve 1.00 MiB using storage policy from min volume index 0
[azat.local] 2023.12.09 17:18:02.834553 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> DiskLocal: Reserved 1.00 MiB on local disk `default`, having unreserved 299.26 GiB.
[azat.local] 2023.12.09 17:18:02.834947 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> MergedBlockOutputStream: filled checksums all_1_1_0 (state Temporary)
[azat.local] 2023.12.09 17:18:02.835050 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Debug> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): Wrote block with ID 'all_8403487165966020028_15352290374794180109', 1 rows
[azat.local] 2023.12.09 17:18:02.835070 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Information> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): commit_new_part_stage
[azat.local] 2023.12.09 17:18:02.835971 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Information> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): lock_part_number_stage 15
[azat.local] 2023.12.09 17:18:02.836006 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77): Renaming temporary part tmp_insert_all_1_1_0 to all_15_15_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
[azat.local] 2023.12.09 17:18:02.836046 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Test> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77): preparePartForCommit: inserting all_15_15_0 (state PreActive) into data_parts_indexes
[azat.local] 2023.12.09 17:18:02.836141 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Debug> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77): Undoing transaction (1, 1, 00000000-0000-0000-0000-000000000000). Removing parts: all_15_15_0.
[azat.local] 2023.12.09 17:18:02.836593 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): ZooKeeperRetriesControl: ReplicatedMergeTreeSink::consume/commitPart: setKeeperError: error=Session expired message=XID overflow
[azat.local] 2023.12.09 17:18:02.836615 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Debug> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): ZooKeeperRetriesControl: ReplicatedMergeTreeSink::consume/commitPart: will retry due to error: retry_count=1 timeout=100ms error=Session expired message=XID overflow
[azat.local] 2023.12.09 17:18:02.936706 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Information> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): commit_new_part_stage
[azat.local] 2023.12.09 17:18:02.937449 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Information> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77) (Replicated OutputStream): lock_part_number_stage 16
[azat.local] 2023.12.09 17:18:02.937489 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Trace> default.data_r1 (66cdd905-a34d-4f0a-a59f-cc4d96dd8e77): Renaming temporary part all_15_15_0 to all_16_16_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
[azat.local] 2023.12.09 17:18:02.967917 [ 11522 ] {551fb116-b0c4-4bcd-82a9-b042b178fa2f} <Error> executeQuery: Code: 230. DB::Exception: Unexpected state of part all_16_16_0 (state Outdated). Expected: Temporary. (NOT_FOUND_EXPECTED_DATA_PART) (version 23.12.1.1) (from [::1]:54026) (in query: insert into data_r1 values ), Stack trace (when copying this message, always include the lines below):

0. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x0000000014a29812 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
1. /src/ch/clickhouse/src/Common/Exception.cpp:97: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000b7337a5 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
2. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<String, String&>(int, FormatStringHelperImpl<std::type_identity<String>::type, std::type_identity<String&>::type>, String&&, String&) @ 0x0000000007a47139 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
3. /src/ch/clickhouse/src/Storages/MergeTree/IMergeTreeDataPart.cpp:579: DB::IMergeTreeDataPart::assertState(std::initializer_list<DB::MergeTreeDataPartState> const&) const @ 0x00000000119cce1b in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
4. /src/ch/clickhouse/src/Storages/MergeTree/MergeTreeData.cpp:3721: DB::MergeTreeData::renameTempPartAndReplaceImpl(std::shared_ptr<DB::IMergeTreeDataPart>&, DB::MergeTreeData::Transaction&, DB::DataPartsLock&, std::vector<std::shared_ptr<DB::IMergeTreeDataPart const>, std::allocator<std::shared_ptr<DB::IMergeTreeDataPart const>>>*) @ 0x0000000011a88320 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
5. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/vector:543: DB::MergeTreeData::renameTempPartAndAdd(std::shared_ptr<DB::IMergeTreeDataPart>&, DB::MergeTreeData::Transaction&, DB::DataPartsLock&) @ 0x0000000011a88c93 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
6. /src/ch/clickhouse/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:0: DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda2'()::operator()() const @ 0x0000000011d88aee in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
7. /src/ch/clickhouse/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:0: DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda3'()::operator()() const @ 0x0000000011d88206 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
8. /src/ch/clickhouse/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:1055: DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda4'()::operator()() const @ 0x0000000011d8806b in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
9. /src/ch/clickhouse/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:1069: void DB::ZooKeeperRetriesControl::retryLoop<DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda4'(), DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda5'()>(DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda4'()&&, DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool)::'lambda5'()&&) @ 0x0000000011d79378 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
10. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/vector:543: DB::ReplicatedMergeTreeSinkImpl<false>::commitPart(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&, std::shared_ptr<DB::IMergeTreeDataPart>&, String const&, unsigned long, bool) @ 0x0000000011d78b97 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
11. /src/ch/clickhouse/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:432: DB::ReplicatedMergeTreeSinkImpl<false>::finishDelayedChunk(std::shared_ptr<DB::ZooKeeperWithFaultInjection> const&) @ 0x0000000011d6d903 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
12. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/vector:434: DB::ReplicatedMergeTreeSinkImpl<false>::consume(DB::Chunk) @ 0x0000000011d77beb in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
13. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::SinkToStorage::onConsume(DB::Chunk) @ 0x000000001232e88b in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
14. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::ExceptionKeepingTransform::work()::$_1, void ()>>(std::__function::__policy_storage const*) @ 0x0000000012273bf9 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
15. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x0000000012273a5f in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
16. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__functional/function.h:818: ? @ 0x00000000122733ff in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
17. /src/ch/clickhouse/src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x00000000120083a9 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
18. /src/ch/clickhouse/src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000011fff810 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
19. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/atomic:958: DB::PipelineExecutor::executeStep(std::atomic<bool>*) @ 0x0000000011fff0f7 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
20. /src/ch/clickhouse/src/Server/TCPHandler.cpp:900: DB::TCPHandler::processInsertQuery() @ 0x0000000011fa1ade in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
21. /src/ch/clickhouse/src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x0000000011f98c81 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
22. /src/ch/clickhouse/src/Server/TCPHandler.cpp:2294: DB::TCPHandler::run() @ 0x0000000011faf1aa in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
23. /src/ch/clickhouse/base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x0000000014937327 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
24. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: Poco::Net::TCPServerDispatcher::run() @ 0x000000001493781a in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
25. /src/ch/clickhouse/base/poco/Foundation/src/ThreadPool.cpp:202: Poco::PooledThread::run() @ 0x0000000014a8577b in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
26. /src/ch/clickhouse/base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000014a82e31 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
27. ? @ 0x00007ffff7e3d9eb in ?
28. ? @ 0x00007ffff7ec17cc in ?

```

Reply:
Please explain it more verbose. What did you add and where?

Reply:
>What did you add and where?

I've added exception before calling `tryMulti`, and I throwed `ZSESSIONEXPIRED` error (you can see it from from the log aboveE) that should be retried, but it failed.

> Once part is tried to be committed there is no way to do retry any more. It is ether committed or rolled back, or committed and after deleted.

It is not committed, the commit failed. The reason is that before your PR the part's state had been rolled back to `Temporary` (in `transaction.rollbackPartsToTemporaryState`), but not it does not, and second `renameTempPartAndAdd` attempt (due to retry) failed because the part's state is `Outdated`

Reply:
If session expires exactly on commit call, that code follows the to the `UNCERTAIN_COMMIT` stage. It is `isHardwareError`. There are no retries, indeed. I did it intentionally. I have strong opinion that we should not broke the flow of part's live cycle.

[Algunenano](https://github.com/Algunenano) suggest more advance handling for `UNCERTAIN_COMMIT` stage. That approach would cover your case.
In short: if we are sure that nothing has been written, we do retry. But part are never going to temporary state agin. It all should happen inside `PreActive` part state.
I like his suggestion and I'm would be glad to help him with that.

Reply:
>There are no retries, indeed. I did it intentionally

But before there was retires for this case, and AFAICS it is the most common error to retry (XID overflow and other session exparations). And plus this PR is marked as `Not for changelog`...

>Part state is only advancing, never goes back

Definitely agree, playing with part state can lead to some tricky cases I guess...

>[Algunenano](https://github.com/Algunenano) suggest more advance handling for UNCERTAIN_COMMIT stage

I guess you are talking about this https://github.com/ClickHouse/ClickHouse/pull/57205#discussion_r1411165127

>In short: if we are sure that nothing has been written, we do retry

Yes, this sounds like will cover most of the cases. Thanks!

Reply:
Actually I think you a wrong. Let me explain. Correct me if I'm wrong.

The old code has the same behaviour.
If connection is lost right before or inside 
https://github.com/CheSema/ClickHouse/blob/bf08f3ef5a54e6fd12db2f5236103b836fb733e1/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp#L924
Than no retries happen. The same consequences.

It is more likely that you are referring to the case when connection is lost here
https://github.com/CheSema/ClickHouse/blob/bf08f3ef5a54e6fd12db2f5236103b836fb733e1/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp#L902C13-L902C65

In that case we had such steps which looked like helping here
https://github.com/CheSema/ClickHouse/blob/bf08f3ef5a54e6fd12db2f5236103b836fb733e1/src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp#L907-L908

But I removed that superfluous call. Now it easier to see that case when no retry happen. It is good actually.

Reply:
>The old code has the same behaviour.

Yes, you are right, I did not look carefully enough to the history. Thanks!
And now I thought about this case, and indeed in case of session loss it cannot be retried blindly.

BTW the XID overflow could be retries, right? But right now there is no separate exception code for it...

```
throw Exception::fromMessage(Error::ZSESSIONEXPIRED, "XID overflow");
                              ^^^^^^^^^^^^^^^^^^^^
```

Reply:
100%. I had the code to introduce both faults and retries here (faults are not introduced in tryMulti currently) in [this PR](https://github.com/ClickHouse/ClickHouse/pull/56917/files#diff-d34dfcef2ee825d0184c61973c0c3c8606d68975870f22c9ddd1501f7ff88e42R925-R969) but then this bug was found and introducing more retries in the same way would have led to more zero-copy nodes lost, so I stopped the PR.

The idea is simple: Recover the ZK connection, if it was written then it's good, if it wasn't written then retry completely (which might require some extra code or recovering some of the removed one).